Fix #3290: Add support for generic regex pattern matching

Sparsh1212 opened the pull request:
## Explanation
Fixes #3290: Add support for generic regex pattern matching

The script traverses different layers of the codebase to check for any prohibited file content and filenaming pattern.
To run the script:
`bazel run //scripts:pattern_validation_check -- $(pwd)`

Tests are added to check for the prohibited file content and filenaming patterns in the `scripts/src/test/testfiles` directory.
To run the tests:
`bazel test //scripts:test_pattern_validation_check`
The tests are designed in a way such that, we analyze the script checks over some dummy testfiles. The testfiles directory has dummy test files present for all the checks under their respective directory. They are further segregated into two directories:

1. `fail`
2. `pass`

`fail` directory contains the files, which will trigger failure for the script checks.
`pass` directory contains the files, which will make the checks pass.

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Sparsh1212 committed:
Fix #3290: Add support for generic regex pattern matching

Sparsh1212 committed:
Apply review suggestions for proto location

Sparsh1212 committed:
Delete main directory

Sparsh1212 committed:
Implment script logic and tests

Sparsh1212 committed:
CI setup part 1

Sparsh1212 committed:
syntax fix in yaml file

Sparsh1212 committed:
import nits in dummy file

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/proto/filename_pattern_validation_checks.proto:
```diff
@@ -0,0 +1,18 @@
+
+  // Error to show if any files match the regex.
+  string failure_message = 2;
+}
```
add eof

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/proto/filename_pattern_validation_checks.proto:
```diff
@@ -0,0 +1,18 @@
+
+  // Error to show if any files match the regex.
+  string failure_message = 2;
+}
```
add eof

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/test/dummyfiles/DummyFile1.kt:
```diff
@@ -0,0 +1 @@
+import android.support.v7.app
```
Add something valid file

anandwana001 replied:
Also add eof

Sparsh1212 replied:
Done regarding eof.
Regarding what content we should add here, I am not sure, as this is just meant to be for the path pattern check.
Adding @BenHenning for additional thoughts.

anandwana001 commented on scripts/assets/file_content_validation_checks.textproto:
```diff
@@ -0,0 +1,5 @@
+  filename_regex: ".+?.kt"
+  prohibited_content_regex: "^import .+?support.+?$"
+  failure_message: "AndroidX should be used instead of the support library"
+}
```
add eof

anandwana001 replied:
Also, we can name something which resemble more on prohibiting the use of support library.

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/RegexPatternValidationCheck.kt:
```diff
@@ -0,0 +1,161 @@
+import org.oppia.android.app.model.FilenameChecks
+import org.oppia.android.app.model.FileContentChecks
+import kotlin.system.exitProcess
+
```
no new line

Sparsh1212 replied:
Shouldn't there be one between the import statements and the class declaration?

anandwana001 replied:
Currently there are 2 empty lines 8th and 9th, so only 1 empty line is enough

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/RegexPatternValidationCheck.kt:
```diff
@@ -0,0 +1,161 @@
+class RegexPatternValidationCheck {
+  companion object {
+    @JvmStatic
+    fun main(vararg args: String) {
```
name should be more descriptive

anandwana001 replied:
also, we don't need to use `JvmStatic`

Sparsh1212 replied:
Actually, here it looks like we need it in order to run the kt_jvm_binary.
I tried removing it, but when running the scripts, it gave me an error, that it wasn't able to recognize the static main method.

anandwana001 replied:
Did you try to change the method name?

Sparsh1212 replied:
The method name must be `main` only. as it is necessary for the `kt_jvm_binary` to run the script. I did this from the reference PR https://github.com/oppia/oppia-android-assets/pull/7/files
In this it is done in the same way.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/RegexPatternValidationCheck.kt:
```diff
@@ -0,0 +1,161 @@
+      val repoPath = args[0] + "/"
+      val allowedDirectories = arrayOf("data", "app")
+      val searchFiles = collectSearchFiles(repoPath, allowedDirectories)
+      var scriptFailedFlag = false
```
we can use more empty line in between

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/RegexPatternValidationCheck.kt:
```diff
@@ -0,0 +1,161 @@
+    ) {
+      println("Prohibited filename pattern: [ROOT]/$filePath")
+      println("Failure message: $errorToShow")
+      println()
```
rather than using extra println, we can use some escape pattern.

anandwana001 replied:
we can use one println only.

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/RegexPatternValidationCheck.kt:
```diff
@@ -0,0 +1,161 @@
+      println()
+    }
+  }
+}
```
add eof

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/RegexPatternValidationCheck.kt:
```diff
@@ -0,0 +1,161 @@
+        println("REGEX PATTERN CHECKS PASSED")
+      }
+    }
+
```
add some kdocs

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/RegexPatternValidationCheck.kt:
```diff
@@ -0,0 +1,161 @@
+      searchFiles.forEach {
+        if (fileNameRegex.matches(it.name)) {
+          var lineNumber = 0
+          File(it.toString()).forEachLine { l ->
```
rather than using `l` single letter naming, use something more descriptve.

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/proto/file_content_validation_checks.proto:
```diff
@@ -0,0 +1,21 @@
+
+    // Error to show if any files matching the filename regex have contents matching the content regex.
+    string failure_message = 3;
+}
```
add eof

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/test/org/oppia/android/scripts/RegexPatternValidationCheckTest.kt:
```diff
@@ -0,0 +1,40 @@
+class RegexPatternValidationCheckTest {
+
+  private val repoPath = System.getProperty("user.dir") + "/"
+  private val testDirectory = arrayOf("scripts/src/test/dummyfiles")
```
use something descriptive naming.

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/test/org/oppia/android/scripts/RegexPatternValidationCheckTest.kt:
```diff
@@ -0,0 +1,40 @@
+  private val searchFiles = RegexPatternValidationCheck.collectSearchFiles(repoPath, testDirectory)
+
+  @Test
+  fun prohibited_Content_Check_Should_Fail() {
```
follow test naming convention

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/test/org/oppia/android/scripts/RegexPatternValidationCheckTest.kt:
```diff
@@ -0,0 +1,40 @@
+  fun prohibited_Content_Check_Should_Fail() {
+    assertEquals(
+      true,
+      RegexPatternValidationCheck.checkProhibitedContent(
```
separate

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/test/org/oppia/android/scripts/RegexPatternValidationCheckTest.kt:
```diff
@@ -0,0 +1,40 @@
+  @Test
+  fun prohibited_Content_Check_Should_Fail() {
+    assertEquals(
+      true,
```
use named param

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/test/org/oppia/android/scripts/RegexPatternValidationCheckTest.kt:
```diff
@@ -0,0 +1,40 @@
+        repoPath = repoPath,
+        searchFiles = searchFiles,
+        fileNameRegexString = "DummyFile1.kt",
+        prohibitedContentRegexString = "^import .+?support.+?$",
```
can we get this string from some single utility.

Sparsh1212 replied:
Updated, to fetch it from the textproto itself.

anandwana001 commented on scripts/src/test/org/oppia/android/scripts/RegexPatternValidationCheckTest.kt:
```diff
@@ -0,0 +1,40 @@
+        searchFiles = searchFiles,
+        fileNameRegexString = "DummyFile1.kt",
+        prohibitedContentRegexString = "^import .+?support.+?$",
+        errorToShow = "AndroidX should be used instead of the support library"
```
same with this.

Sparsh1212 replied:
Done.

anandwana001 commented on scripts/src/test/org/oppia/android/scripts/RegexPatternValidationCheckTest.kt:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  @Test
+  fun no_Prohibited_Content_Check_Should_Pass() {
```
ditto

Sparsh1212 replied:
Done.

anandwana001 reviewed (changes requested):
DummFile2 should be something valid that does not contains any support library.
Also, rather than using a dummy file name, we can use something more descriptive name.  

Add some more description in the PR description on what are the patterns we are testing here. 

Thanks @Sparsh1212

anandwana001 commented:
Removing my assignee as I already reviewed it first time.

Sparsh1212 committed:
Use regex patterns from script in script test

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 commented:
@anandwana001 Thanks for the review,
Addressed all your comments. PTAL.

Sparsh1212 commented:
@fsharpasharp Can you please take a pass for the Bazel files.

Sparsh1212 committed:
Make PR suggestions into effect

fsharpasharp commented on scripts/script_assets.bzl:
```diff
@@ -0,0 +1,80 @@
+
+    Args:
+        name: str. The name of this generation instance. This will be a prefix for derived targets.
+        filename_validation_file_names: list of str. The list of prohibited filename pattern file names.
```
Looks good, my only comment is that the variable names are confusing here. E.g. filename_validation_file_names. Could you update the naming to be a bit clearer especially when combined with the file names:

```
REGEX_PATTERN_CHECK_ASSETS = generate_assets_list_from_text_protos(
    name = "script_assets",
    file_content_validation_file_names = [
        "file_content_validation_checks",
    ],
    filename_validation_file_names = [
        "filename_pattern_validation_checks",
    ],
)
```

Sparsh1212 replied:
Changed it to: `filepath_pattern_validation_file_names`
PTAL, thanks.

Sparsh1212 committed:
Improve naming of script_assets variables

anandwana001 commented:
> @anandwana001 Thanks for the review,
> Addressed all your comments. PTAL.

Assigning you back as there are few comments which I suggested earlier that are not addressed.

fsharpasharp reviewed (approved):
BUILD.bazel file looks good to me.

Sparsh1212 committed:
Make PR suggestions into effect

Sparsh1212 commented:
> > @anandwana001 Thanks for the review,
> > Addressed all your comments. PTAL.
> 
> Assigning you back as there are few comments which I suggested earlier that are not addressed.

Apologies! I missed some of the nits.
Have updated them now. Please let me know if something is still left.

vinitamurthi commented on .github/workflows/static_checks.yml:
```diff
@@ -92,3 +92,18 @@ jobs:
+      - uses: actions/checkout@v2
+
+      - name: Set up Bazel
+        uses: abhinavsingh/setup-bazel@v3
```
Do we want to be using actions that are created by individuals? @BenHenning

Sparsh1212 replied:
We are currently using this action in the unit_tests.yml workflow: https://github.com/oppia/oppia-android/blob/1cc3fc9c9774485219c226c6ccba3e3453513ec5/.github/workflows/unit_tests.yml#L27

vinitamurthi commented on scripts/BUILD.bazel:
```diff
@@ -0,0 +1,42 @@
+kt_jvm_library(
+    name = "pattern_validation_lib",
+    testonly = True,
+    srcs = glob(["src/java/org/oppia/android/scripts/*.kt"]),
```
Can we avoid using glob? In general its better to explicitly list down the files rather than giving a generic rule like this

Sparsh1212 replied:
Done.

vinitamurthi commented on scripts/BUILD.bazel:
```diff
@@ -0,0 +1,42 @@
+kt_jvm_test(
+    name = "test_pattern_validation_check",
+    testonly = True,
+    srcs = glob(["src/test/org/oppia/android/scripts/RegexPatternValidationCheckTest.kt"]),
```
Same comment here and below

Sparsh1212 replied:
Done.

vinitamurthi commented on scripts/src/java/org/oppia/android/scripts/proto/BUILD.bazel:
```diff
@@ -0,0 +1,26 @@
+proto_library(
+    name = "filename_pattern_validation_structure_proto",
+    srcs = ["filename_pattern_validation_checks.proto"],
+    visibility = ["//visibility:public"],
```
Do we need visibility to be public? Again, we like to avoid using public as much as possible. Is it possible to provide a set of subpackages for visibility here and below?

Sparsh1212 replied:
Done.

vinitamurthi commented on scripts/src/java/org/oppia/android/scripts/proto/file_content_validation_checks.proto:
```diff
@@ -0,0 +1,21 @@
+option java_multiple_files = true;
+
+message FileContentChecks {
+  repeated FileContentCheck file_content_checks = 1;
```
Add a comment explaining the field

Sparsh1212 replied:
Done.

vinitamurthi commented on scripts/src/java/org/oppia/android/scripts/proto/file_content_validation_checks.proto:
```diff
@@ -0,0 +1,21 @@
+option java_package = "org.oppia.android.app.model";
+option java_multiple_files = true;
+
+message FileContentChecks {
```
Add a comment explaining the message

Sparsh1212 replied:
Done.

vinitamurthi commented on scripts/src/java/org/oppia/android/scripts/proto/file_content_validation_checks.proto:
```diff
@@ -0,0 +1,21 @@
+  repeated FileContentCheck file_content_checks = 1;
+}
+
+message FileContentCheck {
```
Add a comment explaining the message

Sparsh1212 replied:
Done.

vinitamurthi commented on scripts/src/java/org/oppia/android/scripts/proto/filename_pattern_validation_checks.proto:
```diff
@@ -0,0 +1,18 @@
+option java_package = "org.oppia.android.app.model";
+option java_multiple_files = true;
+
+message FilenameChecks {
```
Add comments here and below

Sparsh1212 replied:
Done.

Sparsh1212 committed:
Make PR suggestions

Sparsh1212 committed:
Make PR suggestions

Sparsh1212 commented:
@vinitamurthi Addressed all your suggestions. Please take a look.

Sparsh1212 commented:
@anandwana001 @vinitamurthi @fsharpasharp @BenHenning 
Apologies! had to close this one, as this was from a fork, so the upcoming stacked PRs cannot be chained on this one.
All the review comments have been incorporated in the new PR #3340
